### PR TITLE
Player: Journal funcs

### LINKED ISSUE #807: AddCustomJournalEntry  - to player plugin
Just placeholder issue to link against

Currently building method to add custom journal entries to a player via NSS / C#
Without having to use Custom Tokens or modify the JRL file.

### PR DESCRIPTION
**Pull request by Baaleos**
Journal modification functionality
As demonstrated here:
https://youtu.be/ZeR-9S6WAwc

resolves #807

### CODE REVIEW

File: Plugins/Player/Player.cpp [hunk 1]
```diff
@@ -1336,4 +1344,171 @@ ArgumentStack Player::SetResManOverride(ArgumentStack&& args)
     return Services::Events::Arguments();
 }
 
+
+CExoLocString CreateCExoLocString(const std::string& str)
+{
+    CExoLocString locStr;
+    locStr.AddString(0,CExoString(str.c_str()),0);
+    return locStr;
+}
```
Comment (Daztek) [inline Plugins/Player/Player.cpp]: This should go in NWNXLib/Utils.hpp/cpp as the following:

```c
CExoLocString CreateLocString(const std::string& str, int32_t nID = 0, uint8_t bGender = 0);

CExoLocString CreateLocString(const std::string& str, int32_t nID, uint8_t bGender)
{
    CExoLocString locStr;
    locStr.AddString(nID, CExoString(str.c_str()), bGender);
    
    return locStr;
}
```

File: Plugins/Player/Player.cpp [hunk 2]
```diff
@@ -1336,4 +1344,171 @@ ArgumentStack Player::SetResManOverride(ArgumentStack&& args)
...
+            auto entries = pCreature->m_pJournal->m_lstEntries;
+            
+            const auto questName = Services::Events::ExtractArgument<std::string>(args);
+            const auto questText = Services::Events::ExtractArgument<std::string>(args);
+            const auto tag = Services::Events::ExtractArgument<std::string>(args);
+            const auto state = Services::Events::ExtractArgument<int32_t>(args);
+            const auto priority = Services::Events::ExtractArgument<int32_t>(args);
+            const auto completed = Services::Events::ExtractArgument<int32_t>(args);
+            const auto diplayed = Services::Events::ExtractArgument<int32_t>(args);
+            const auto updated = Services::Events::ExtractArgument<int32_t>(args);
```
Comment (Daztek) [inline Plugins/Player/Player.cpp]: Some `ASSERT_OR_THROW()`s would be nice here. Especially for the `SJournalEntry` stuff that takes an unsigned value.

Also have you tested what happens when `questName`, `questText` or `tag` are empty strings?

File: Plugins/Player/Player.cpp [hunk 3]
```diff
@@ -1336,4 +1344,171 @@ ArgumentStack Player::SetResManOverride(ArgumentStack&& args)
...
+            auto silentUpdate = Services::Events::ExtractArgument<int32_t>(args);
+            
+            //If server owner leaves this 0 - the entry will be added with todays date
+            if (calDay == 0)
+            {
+                uint32_t year = Globals::AppManager()->m_pServerExoApp->GetWorldTimer()->GetWorldTimeYear();
+                uint32_t month = Globals::AppManager()->m_pServerExoApp->GetWorldTimer()->GetWorldTimeMonth();
+                uint32_t day = Globals::AppManager()->m_pServerExoApp->GetWorldTimer()->GetWorldTimeDay();
+                calDay = Globals::AppManager()->m_pServerExoApp->GetWorldTimer()->ConvertToCalendarDay(year, month, day);
+            }
```
Comment (Daztek) [inline Plugins/Player/Player.cpp]: I think you can just do `calDay = Globals::AppManager()->m_pServerExoApp->GetWorldTimer()->GetWorldTimeCalendarDay();`

File: Plugins/Player/Player.cpp [hunk 4]
```diff
@@ -1336,4 +1344,171 @@ ArgumentStack Player::SetResManOverride(ArgumentStack&& args)
...
+            //If server owner leaves this 0 - the entry will be added with now() time
+            if(timeDay == 0)
+            {
+                uint32_t hour = Globals::AppManager()->m_pServerExoApp->GetWorldTimer()->GetWorldTimeHour();
+                uint32_t minute = Globals::AppManager()->m_pServerExoApp->GetWorldTimer()->GetWorldTimeMinute();
+                uint32_t second = Globals::AppManager()->m_pServerExoApp->GetWorldTimer()->GetWorldTimeSecond();
+                uint32_t ms = Globals::AppManager()->m_pServerExoApp->GetWorldTimer()->GetWorldTimeMillisecond();
+                timeDay = Globals::AppManager()->m_pServerExoApp->GetWorldTimer()->ConvertToTimeOfDay(hour, minute, second, ms);
+            }
```
Comment (Daztek) [inline Plugins/Player/Player.cpp]: And `timeDay = Globals::AppManager()->m_pServerExoApp->GetWorldTimer()->GetWorldTimeTimeOfDay();` here

File: Plugins/Player/Player.cpp (hunk 1 above)
Comment (Baaleos) [reply Plugins/Player/Player.cpp]: Have moved it to Utils as requested

File: Plugins/Player/Player.cpp (hunk 2 above)
Comment (Baaleos) [reply Plugins/Player/Player.cpp]: Added ASSERT OR THROW on the tag - my tests showed that the tag is the only parameter that is actually required for success.

File: Plugins/Player/Player.cpp (hunk 3 above)
Comment (Baaleos) [reply Plugins/Player/Player.cpp]: Done

File: Plugins/Player/Player.cpp (hunk 4 above)
Comment (Baaleos) [reply Plugins/Player/Player.cpp]: Done

File: Plugins/Player/Player.cpp (hunk 2 above)
Comment (Daztek) [reply Plugins/Player/Player.cpp]:
```c
    uint32_t nCalendarDay;
    uint32_t nTimeOfDay;
    uint32_t nState;
    uint32_t nPriority;
    uint16_t nPictureIndex;
```

These are all unsigned, should have an `ASSERT_OR_THROW(var >= 0)`

Comment (Baaleos) [reply Plugins/Player/Player.cpp]: added extra combined assert or throw-  now they are all checked to be >= 0

Comment (Daztek) [reply Plugins/Player/Player.cpp]: Haha, sorry D: but we prefer them not combined so it's easier to see which assert is being thrown.

Comment (Baaleos) [reply Plugins/Player/Player.cpp]: lol - done
Reason I kept the 'tag' assertion higher up - is because it is the bare minimum required for the method to function

Comment (Baaleos) [reply Plugins/Player/Player.cpp]: actually done now: not sure what happened last night - must have forgot to push my changes

File: Plugins/Player/Player.cpp [hunk 5]
```diff
@@ -1336,4 +1344,167 @@ ArgumentStack Player::SetResManOverride(ArgumentStack&& args)
...
+            newJournal.nPictureIndex= 0; // Not implemented by bioware/beamdog  
+            newJournal.bQuestCompleted= completed; 
+            newJournal.bQuestDisplayed= diplayed; 
+            newJournal.bUpdated     = updated; 
+                   
+            auto *pMessage = static_cast<CNWSMessage*>(Globals::AppManager()->m_pServerExoApp->GetNWSMessage());       
+            if (pMessage)
+                {
+                 
+                    int overwrite = -1;
```
Comment (GoLoT) [inline Plugins/Player/Player.cpp]: It looks like you are using overwrite as a bool instead of an index, the `overwrite = i;` isn't used. I'd rewrite the loop as:

``` cpp
int i = 0;
for (i = 0; i < entries.num; i++)
{
    auto* pEntry = entries.element[i];
    if (pEntry->szPlot_Id.CStr() == tag)
    {
        pCreature->m_pJournal->m_lstEntries[i] = newJournal;
        break;
    }
}
if(i == entries.num)
    pCreature->m_pJournal->m_lstEntries.Add(newJournal);

```

Comment (Baaleos) [reply Plugins/Player/Player.cpp]: Had some issues implementing the approach above 'exactly'
Merged the main principle described above into the existing for loop structure.

File: Plugins/Player/Player.cpp [hunk 6]
```diff
@@ -1336,4 +1344,163 @@ ArgumentStack Player::SetResManOverride(ArgumentStack&& args)
...
+        auto *pCreature = Globals::AppManager()->m_pServerExoApp->GetCreatureByGameObjectID(pPlayer->m_oidNWSObject);
+        if (pCreature && pCreature->m_pJournal)
+        {
+            auto entries = pCreature->m_pJournal->m_lstEntries;
+            const auto tag = Services::Events::ExtractArgument<std::string>(args);
+            ASSERT_OR_THROW(tag.size() > 0);
+            if (entries.num > 0)
+            {
+                auto pEntry = entries.element;
+                for (int i = 0; i < entries.num; i++, pEntry++)
```
Comment (mtijanic) [inline Plugins/Player/Player.cpp]: Please no pointer arithmetic :)

`CExoArrayList<>` actually supports modern C++ iterators, so you can probably do:
```C
for (auto& entry : entries)
{
  ...
}
```

File: Plugins/Player/Player.cpp [hunk 7]
```diff
@@ -1336,4 +1344,163 @@ ArgumentStack Player::SetResManOverride(ArgumentStack&& args)
...
+                {
+                    if (pEntry->szPlot_Id.CStr() == tag)
+                        {
+                            lastJournalEntry = entries[i];
+                            found = i;
```
Comment (mtijanic) [inline Plugins/Player/Player.cpp]: Should we abort if we find the tag, or do we need to look at the last entry with the same tag?
If there's only one with the tag, maybe move the big return here and then you don't need the `found` variable at all.

File: Plugins/Player/Player.cpp [hunk 8]
```diff
@@ -1336,4 +1344,163 @@ ArgumentStack Player::SetResManOverride(ArgumentStack&& args)
...
+            ASSERT_OR_THROW(silentUpdate >= 0);
+            
+            
+            //If server owner leaves this 0 - the entry will be added with todays date
+            if (calDay <= 0)
+            {
+                calDay = Globals::AppManager()->m_pServerExoApp->GetWorldTimer()->GetWorldTimeCalendarDay();
+            }
+            //If server owner leaves this 0 - the entry will be added with now() time
+            if(timeDay <= 0)
```
Comment (mtijanic) [inline Plugins/Player/Player.cpp]: A few generic style nits throughout the change:
- space between keyword and paren: `if (x)`, not `if(x)`.
- space around assignment operator `=` on both sides:  `a = x;` not `a=x`, `a= x`, `a =x`.
- Avoid superfluous newlines, such as empty lines after `{`

File: Plugins/Player/Player.cpp (hunk 7 above)
Comment (Baaleos) [reply Plugins/Player/Player.cpp]: Default journal entries add multiple entries to the player with the same tag - the last one is (as far as I can tell) the one that is generally presented/visible to the player. Even when you remove a journal entry via native RemoveJournalEntry NWScript method - it only ever removes the last entry - leaving some empty journal entries in the journal list on the player. (Somewhat memory-leakish behaviour in the base game)

The AddCustomJournal entry method attempts to avoid that sort of behaviour by overwriting entries with the same tag. This Get method should be compatible with both native journal entries and the custom entries.

Also - the found variable is used so I can return some sort of indication to the NWScript context that no valid journal entry was found - eg: no need to do more popping of return variables if the journal entry was not found etc.
Seemed more efficient to pop one return variable with -1 to indicate to the nwscript that there is no further variables to be expected etc.
If there is a better way of handling that - happy to include it.

Comment (mtijanic) [reply Plugins/Player/Player.cpp]: In that case, it would be better to start at the last element and iterate backwards until you find the first tag.

And what I mean is if you add the full data return here, you don't need the `found` variable. As soon as you hit the end of the function, you know it's not found, so you can just return `-1` directly.

Comment (Baaleos) [reply Plugins/Player/Player.cpp]: Ok - will try to refactor to start at entries.num -1 and work down towards 0 etc
Will also add the return directly to the loop

**Comment by mtijanic** [comment]
Is this ready for merge? (i didn't re-eeview yet since last time)

**Comment by Baaleos** [comment]
> Is this ready for merge? (i didn't re-eeview yet since last time)

Not quite - It still needs to have the style nits fixed and needs to be retested against the latest server binary. I will try to get some time put aside for this at the weekend.

**Comment by plenarius** [comment]
Reviving this as I took my time to do my own review and fixed the merge conflicts and wouldn't mind seeing this added. Will spend some time testing a bit.

**Comment by Baaleos** [comment]
> Reviving this as I took my time to do my own review and fixed the merge conflicts and wouldn't mind seeing this added. Will spend some time testing a bit.

Happy for someone to take it over - I've been pre-occupied with RL stuff at the moment.
From what I recall - everything was working - but then I implemented some of the refactoring that was recommended, such as iterating through the list backwards etc for efficiency. Just never got a chance to test that part etc.
